codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,467 @@
+            // We return the bytes in bytesConsumed() once the stream channel consumed the bytes.
+            return 0;
+        }
+    }
+
+    static final class StreamInfo {
+        final Http2StreamChannel childChannel;
+        /**
+         * True if stream is in {@link Http2MultiplexCodec#steamsToFireChildReadComplete}.
+         */
+        boolean inStreamsToFireChildReadComplete;
+
+        StreamInfo(Http2StreamChannel childChannel) {
+            this.childChannel = childChannel;
+        }
+    }
+
+    // This class uses invoker(ctx).invoke* instead of ctx.* to send to the ctx's handler instead
+    // of the 'next' handler.
+    class Http2StreamChannel extends AbstractHttp2StreamChannel {
+        private final Http2Stream stream;
+        boolean onStreamClosedFired;
+
+        Http2StreamChannel(Http2Stream stream) {
+            super(ctx.channel());
+            this.stream = stream;
+        }
+
+        @Override
+        protected void doClose() throws Exception {
+            if (!onStreamClosedFired) {
+                Http2StreamFrame resetFrame = new DefaultHttp2ResetFrame(Http2Error.CANCEL).setStream(this);
+                invoker(ctx).invokeWrite(ctx, resetFrame, ctx.newPromise());
+                invoker(ctx).invokeFlush(ctx);
+            }
+            super.doClose();
+        }
+
+        @Override
+        protected void doWrite(Object msg) {
+            if (!(msg instanceof Http2StreamFrame)) {
+                throw new IllegalArgumentException("Message must be an Http2StreamFrame:" + msg);
+            }
+            Http2StreamFrame frame = (Http2StreamFrame) msg;
+            if (frame.stream() != null) {
+                ReferenceCountUtil.release(frame);
+                throw new IllegalArgumentException("Stream must be null on the frame");
+            }
+            frame.setStream(this);
+            invoker(ctx).invokeWrite(ctx, frame, ctx.newPromise());
+        }
+
+        @Override
+        protected void doWriteComplete() {
+            invoker(ctx).invokeFlush(ctx);
+        }
+
+        @Override
+        protected EventExecutor preferredEventExecutor() {
+            return ctx.executor();
+        }
+
+        @Override
+        protected void bytesConsumed(final int bytes) {
+            if (ctx.executor().inEventLoop()) {
+                bytesConsumed0(bytes);
+            } else {
+                ctx.executor().execute(new OneTimeTask() {
+                    @Override
+                    public void run() {
+                        bytesConsumed0(bytes);
+                    }
+                });
+            }
+        }
+
+        private void bytesConsumed0(int bytes) {
+            try {
+                http2Handler.connection().local().flowController().consumeBytes(stream, bytes);
+            } catch (Throwable t) {
+                invoker(ctx).invokeExceptionCaught(ctx, t);
```

Comment:
It is fired to the parent's pipeline (`ctx` is the parent's context). I used the invoker here because I needed to call the handler directly and wanted to avoid yet-another try-catch within the try catch.

I've replaced it to a call to `ChannelHandlerInvokerUtil.invokeExceptionCaughtNow()`